app/src/physical_layouts.c

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2024 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <zephyr/devicetree.h>
+#include <zephyr/device.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <zephyr/drivers/kscan.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#include <zmk/physical_layouts.h>
+#include <zmk/event_manager.h>
+#include <zmk/events/position_state_changed.h>
+
+#define DT_DRV_COMPAT zmk_physical_layout
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+#define ZKPA_INIT(i, n)                                                                            \
+    {                                                                                              \
+        .width = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, width),                          \
+        .height = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, height),                        \
+        .x = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, x),                                  \
+        .y = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, y),                                  \
+        .rx = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, rx),                                \
+        .ry = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, ry),                                \
+        .r = (int16_t)(int32_t)DT_INST_PHA_BY_IDX(n, keys, i, r),                                  \
+    }
+
+#define ZMK_LAYOUT_INST(n)                                                                         \
+    static const struct zmk_key_physical_attrs _CONCAT(_zmk_physical_layout_keys_,                 \
+                                                       n)[DT_INST_PROP_LEN_OR(n, keys, 0)] = {     \
+                                                                                                   \
+        LISTIFY(DT_INST_PROP_LEN_OR(n, keys, 0), ZKPA_INIT, (, ), n)};                             \
+    ZMK_MATRIX_TRANSFORM_EXTERN(DT_INST_PHANDLE(n, transform));                                    \
+    static const struct zmk_physical_layout _CONCAT(_zmk_physical_layout_, DT_DRV_INST(n)) = {     \
+        .display_name = DT_INST_PROP_OR(n, display_name, "Layout #" #n),                           \
+        .matrix_transform = ZMK_MATRIX_TRANSFORM_T_FOR_NODE(DT_INST_PHANDLE(n, transform)),        \
+        COND_CODE_1(DT_INST_PROP_LEN(n, kscan),                                                    \
+                    (.kscan = DEVICE_DT_GET(DT_INST_PHANDLE(n, kscan)), ), ())                     \
+            .keys = _CONCAT(_zmk_physical_layout_keys_, n),                                        \
```

Comment:
Some indentation funkiness here?

Reply:
`clang-format` butchered this. Let me see if there's a nicer way that it will also be happy with.